### PR TITLE
Add Wien2k

### PR DESCRIPTION
**Pull request by saurabh02**
@maxhutch this PR includes almost all of the format that I will use for the remaining files. I wanted to get this out to start the first round of review so I can add the rest of repeating code later without having to edit a lot more code. 

To note in this PR:- the methods `get_composition`, `get_output_structure` are not implemented for this parser as we have not been provided with the relevant files to gather this info from. Suggestions are welcome for this.

### CODE REVIEW

**Comment by saurabh02** [comment]
I will look into the test logs

**Comment by saurabh02** [comment]
@maxhutch okay, ready now

File: dfttopif/parsers/base.py [hunk 1]
```diff
@@ -179,137 +194,165 @@ def get_positions(self):
...
 
     def get_total_force(self):
-       return None
+        return None
 
     def get_total_magnetization(self):
-       return None
+        return None
+
+    def get_absorption(self):
```
Comment (maxhutch) [inline dfttopif/parsers/base.py]: If this only applies to `Wien2k`, why don't we leave it in that parser and overload `get_result_functions` instead.

**Comment by maxhutch** [comment]
@saurabh02 to deploy, this will also need a `description_wien2k.yaml` file.

File: dfttopif/parsers/wien2k.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        if len(matches) == 0:
+            return None
+        band_gap = matches[-1]["band gap"]
+        band_gap_units = matches[-1]["band gap units"]
+        return Property(scalars=[Scalar(value=band_gap)], units=band_gap_units)
+
+    @staticmethod
+    def _get_wavelengths(energy_lst):
+        wavelgths_lst = []
+        for energy in energy_lst:
```
Comment (maxhutch) [inline dfttopif/parsers/wien2k.py]: This is just begging to be a list comprehension.

File: dfttopif/parsers/wien2k.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+            wavelgths_lst.append(Scalar(value=energy/1240))
+
+        return wavelgths_lst
+
+    @staticmethod
+    def _get_scalars_lst(floats_lst):
+        scalar_objs_lst = []
+        for num in floats_lst:
```
Comment (maxhutch) [inline dfttopif/parsers/wien2k.py]: This is just begging to be a list comprehension.

File: dfttopif/parsers/wien2k.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        # Get data from the .absorp file
+        for filename in os.listdir(directory):
+            if os.path.splitext(filename)[1] == ".absorp":
+                file_path = os.path.join(directory, filename)
+        if not file_path:
+            return None
+
+        parser = AbsorpParser()
+        with open(file_path, "r") as fp:
+            matches = list(filter(lambda x: "energy" in x, parser.parse(fp.readlines())))
```
Comment (maxhutch) [inline dfttopif/parsers/wien2k.py]: This is just begging to be a list comprehension.

File: dfttopif/parsers/wien2k.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        # Get data the .scf file
+        for filename in os.listdir(self._directory):
+            if os.path.splitext(filename)[1] == ".scf":
+                file_path = os.path.join(self._directory, filename)
+        if not file_path:
+            return None
+
+        parser = ScfParser()
+        with open(file_path, "r") as fp:
+            matches = list(filter(lambda x: "total energy" in x, parser.parse(fp.readlines())))
```
Comment (maxhutch) [inline dfttopif/parsers/wien2k.py]: This is just begging to be a list comprehension.

File: dfttopif/parsers/base.py [hunk 6]
```diff
@@ -57,67 +61,75 @@ def __init__(self, directory):
...
+            'Band Gap Energy': 'get_band_gap',
+            'Pressure': 'get_pressure',
+            'Density of States': 'get_dos',
             'Positions': 'get_positions',
             'Forces': 'get_forces',
             'Total force': 'get_total_force',
             'Density': 'get_density',
             'OUTCAR': 'get_outcar',
             'Total magnetization': 'get_total_magnetization',
+            'Optical conductivity xx (Re $\sigma_{xx}$)': 'get_optical_conductivity_xx',
```
Comment (maxhutch) [inline dfttopif/parsers/base.py]: Why add these here if they are wien2k specific?

File: dfttopif/parsers/pwscf.py [hunk 7]
```diff
@@ -390,3 +446,19 @@ def get_xc_energy_contribution(self):
 
     def get_ewald_energy_contribution(self):
         return self._get_key_with_units("ewald energy contribution")
+
+    @staticmethod
+    def get_optical_conductivity_xx():
```
Comment (maxhutch) [inline dfttopif/parsers/pwscf.py]: Why do these need to be here?

File: dfttopif/parsers/vasp.py [hunk 8]
```diff
@@ -369,3 +398,19 @@ def get_total_magnetization(self):
             return None
         total_magnetization = matches[-1]["total magnetization"]
         return Property(scalars=[Scalar(value=total_magnetization)], units="Bohr")
+
+    @staticmethod
+    def get_optical_conductivity_xx():
```
Comment (maxhutch) [inline dfttopif/parsers/vasp.py]: Why do these need to be here?

File: dfttopif/parsers/wien2k.py [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+                with open(os.path.join(self._directory, filename)) as fp:
+
+                    # look for line with ":LABEL3:"
+                    for line in fp:
+                        if ":LABEL3:" in line:
+                            words = line.split()
+                            return words[2].strip("WIEN2k_")
+
+                # Error handling: version not found
+                raise Exception("Wien2k version not found")
```
Comment (maxhutch) [inline dfttopif/parsers/wien2k.py]: Which kind of exception?  `Exception` is too vague.

File: dfttopif/parsers/wien2k.py [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    """
+    Parser for Wien2k calculations
+    """
+
+    def get_name(self): return "Wien2k"
+    
+    def test_if_from(self, directory):
+        # Check whether it has a .scf file (analogous to an OUTCAR file)
+        for filename in os.listdir(directory):
+            if os.path.splitext(filename)[1] == ".scf":
```
Comment (maxhutch) [inline dfttopif/parsers/wien2k.py]: This could match some pwscf files.  Could you inspect the `scf` file for some `wien2k`-only info?

**Comment by saurabh02** [comment]
@maxhutch ready to merge and deploy now

**Pull request merged**